File: secure-gateway-studio/backend/src/sgstudio/providers/observability.py
from __future__ import annotations

import json
from datetime import UTC, datetime, timedelta
from typing import Any

from google.auth.exceptions import DefaultCredentialsError, RefreshError

from sgstudio.domain.models import (
    BackendKind,
    DeploymentSpec,
    GatewayLogCategory,
    GatewayLogEntry,
    GatewayLogsResponse,
)
from sgstudio.providers.google_rest import GoogleAuthorizedTransport, JsonTransport

LOG_FIELDS = (
    "entries(insertId,timestamp,severity,protoPayload(methodName,status),"
    "jsonPayload(request_id,requestId,role,method,status,outcome,decision,"
    "request_time,tls_protocol,upstream_status,upstream_time)),nextPageToken"
)
MAX_LOG_PAGES = 100
OPERATIONAL_PAYLOAD_KEYS = (
    "role",
    "method",
    "status",
    "outcome",
    "decision",
    "request_time",
    "tls_protocol",
    "upstream_status",
    "upstream_time",
)


class GoogleGatewayObservability:
    def __init__(self, transport: JsonTransport) -> None:
        self._transport = transport

    def list_logs(
        self,
        spec: DeploymentSpec,
        *,
        run_id: str,
        category: GatewayLogCategory,
        hours: int = 24,
        limit: int = 100,
    ) -> GatewayLogsResponse:
        safe_hours = max(1, min(hours, 168))
        safe_limit = max(1, min(limit, 200))
        since = datetime.now(UTC) - timedelta(hours=safe_hours)
        filters = [f'timestamp>="{since.isoformat()}"']
        setup_notice: str | None = None
        data_access_notice = False
        logging_enabled: bool | None = None

        if category is GatewayLogCategory.ACCESS:
            filters.extend(
                [
                    'resource.type="audited_resource"',
                    'resource.labels.method="AuthorizeUser"',
                    'resource.labels.service="beyondcorp.googleapis.com"',
                ]
            )
            data_access_notice = True
        elif category is GatewayLogCategory.CONNECTION:
            logging_enabled = self._gateway_logging_enabled(spec)
            filters.append('resource.type="beyondcorp.googleapis.com/SecurityGateway"')
        elif category is GatewayLogCategory.ADMIN:
            filters.extend(
                [
                    'resource.type="audited_resource"',
                    'resource.labels.service="beyondcorp.googleapis.com"',
                ]
            )
        else:
            instance_id = self._offload_instance_id(spec)
            if instance_id is None:
                return GatewayLogsResponse(
                    run_id=run_id,
                    category=category,
                    entries=[],
                    logging_enabled=logging_enabled,
                    setup_notice=(
                        "Nginx logs require an HTTP-offload VM and the Google Cloud "
                        "Ops Agent. This architecture has no Nginx tier."
                        if spec.backend_kind
                        in {BackendKind.DIRECT_HTTPS, BackendKind.INTERNAL_HTTPS_LB}
                        else "The offload VM was not found or its instance ID is unavailable."
                    ),
                )
            filters.extend(
                [
                    'resource.type="gce_instance"',
                    f'resource.labels.instance_id="{instance_id}"',
                    '(log_id("nginx_access") OR log_id("nginx_access.log"))',
                ]
            )
            setup_notice = (
                "Nginx entries appear after the Google Cloud Ops Agent is configured "
                "to collect /var/log/nginx/sgstudio-access.log."
            )

        raw_entries: list[dict[str, Any]] = []
        seen_page_tokens: set[str] = set()
        page_token: str | None = None
        for page in range(MAX_LOG_PAGES):
            body: dict[str, Any] = {
                "resourceNames": [f"projects/{spec.project_id}"],
                "filter": " AND ".join(filters),
                "orderBy": "timestamp desc",
                "pageSize": min(safe_limit - len(raw_entries), 200),
            }
            if page_token is not None:
                body["pageToken"] = page_token
            _, payload = self._transport.request_json(
                "POST",
                "https://logging.googleapis.com/v2/entries:list",
                params={"fields": LOG_FIELDS},
                json_body=body,
            )
            page_entries = payload.get("entries", [])
            if not isinstance(page_entries, list) or any(
                not isinstance(item, dict) for item in page_entries
            ):
                raise ValueError("cloud-logging-entries-invalid")
            raw_entries.extend(page_entries[: safe_limit - len(raw_entries)])

            if "nextPageToken" not in payload or payload["nextPageToken"] == "":
                break
            next_page_token = payload["nextPageToken"]
            if (
                not isinstance(next_page_token, str)
                or next_page_token in seen_page_tokens
            ):
                raise ValueError("cloud-logging-page-token-invalid")
            seen_page_tokens.add(next_page_token)
            if len(raw_entries) >= safe_limit:
                break
            if page + 1 >= MAX_LOG_PAGES:
                raise ValueError("cloud-logging-pagination-incomplete")
            page_token = next_page_token

        entries = [
            self._entry(item, category, index)
            for index, item in enumerate(raw_entries)
        ]
        return GatewayLogsResponse(
            run_id=run_id,
            category=category,
            entries=entries,
            logging_enabled=logging_enabled,
            data_access_notice=data_access_notice,
            setup_notice=setup_notice,
        )

    def _gateway_logging_enabled(self, spec: DeploymentSpec) -> bool:
        # The empty LoggingConfig message is the connection-logging enable
        # marker. A field mask keeps this state read from receiving unrelated
        # Security Gateway provider output.
        _, payload = self._transport.request_json(
            "GET",
            (
                "https://beyondcorp.googleapis.com/v1/"
                f"projects/{spec.project_id}/locations/global/"
                f"securityGateways/{spec.gateway_id}"
            ),
            params={"fields": "logging"},
        )
        if "logging" not in payload:
            return False
        logging = payload["logging"]
        if not isinstance(logging, dict) or logging:
            raise ValueError("security-gateway-logging-state-invalid")
        return True

    def _offload_instance_id(self, spec: DeploymentSpec) -> str | None:
        if spec.backend_kind in {
            BackendKind.DIRECT_HTTPS,
            BackendKind.INTERNAL_HTTPS_LB,
        }:
            return None
        if spec.mode.value != "poc":
            return None
        url = (
            f"https://compute.googleapis.com/compute/v1/projects/{spec.project_id}/"
            f"zones/{spec.zone}/instances/{spec.name}-offload"
        )
        status, payload = self._transport.request_json(
            "GET", url, accepted_statuses=(200, 404)
        )
        if status == 404:
            return None
        identifier = payload.get("id")
        return str(identifier) if isinstance(identifier, (str, int)) else None

    @classmethod
    def _entry(
        cls,
        item: dict[str, Any],
        category: GatewayLogCategory,
        index: int,
    ) -> GatewayLogEntry:
        proto = item.get("protoPayload") if isinstance(item.get("protoPayload"), dict) else {}
        json_payload = (
            item.get("jsonPayload") if isinstance(item.get("jsonPayload"), dict) else {}
        )
        method = proto.get("methodName")
        request_id = json_payload.get("request_id") or json_payload.get("requestId")
        summary = cls._summary(proto, json_payload)
        timestamp = item.get("timestamp")
        parsed_timestamp: datetime | None = None
        if isinstance(timestamp, str):
            try:
                parsed_timestamp = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
            except ValueError:
                parsed_timestamp = None
        safe_payload = cls._operational_payload(proto, json_payload)
        return GatewayLogEntry(
            insert_id=str(item.get("insertId") or f"entry-{index}"),
            timestamp=parsed_timestamp,
            severity=str(item.get("severity") or "DEFAULT")[:32],
            category=category,
            summary=summary,
            principal=None,
            method=str(method)[:300] if isinstance(method, str) else None,
            resource=None,
            request_id=str(request_id)[:200] if isinstance(request_id, str) else None,
            payload=safe_payload,
        )

    @staticmethod
    def _summary(
        proto: dict[str, Any], json_payload: dict[str, Any]
    ) -> str:
        for key in ("status", "outcome", "decision"):
            value = json_payload.get(key)
            if isinstance(value, (str, int, bool)):
                return str(value)[:500]
        status = proto.get("status")
        if isinstance(status, dict) and status:
            return json.dumps(status, ensure_ascii=False, sort_keys=True)[:500]
        method = proto.get("methodName")
        return str(method)[:500] if method else "Secure Gateway log entry"

    @staticmethod
    def _operational_payload(
        proto: dict[str, Any], json_payload: dict[str, Any]
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {}
        status = proto.get("status")
        if isinstance(status, dict) and isinstance(status.get("code"), int):
            payload["api_status_code"] = status["code"]
        for key in OPERATIONAL_PAYLOAD_KEYS:
            value = json_payload.get(key)
            if isinstance(value, str):
                payload[key] = value[:200]
            elif isinstance(value, (int, float, bool)):
                payload[key] = value
        return payload


def create_google_gateway_observability(
    *, require_impersonation: bool = False
) -> GoogleGatewayObservability:
    try:
        transport = GoogleAuthorizedTransport.from_adc(
            require_impersonation=require_impersonation
        )
    except (DefaultCredentialsError, RefreshError) as error:
        raise RuntimeError("Application Default Credentials are unavailable for logs.") from error
    return GoogleGatewayObservability(transport)

File: secure-gateway-studio/extension/src/providers/observability.ts
/**
 * Post-deployment log views. Port of `providers/observability.py`.
 *
 * Four sanitized views over Cloud Logging, each a fixed filter rather than a
 * free-text query. The operator cannot compose their own filter, which is the
 * point: a query box over an administrator's project is a data-exfiltration
 * surface, and the four questions worth asking after a deployment are known in
 * advance.
 */

import type { DeploymentSpec } from "../domain/spec.ts";
import type { Transport } from "./executor.ts";

export type LogCategory = "access" | "connection" | "admin" | "nginx";

export interface LogEntry {
  insert_id: string;
  category: LogCategory;
  timestamp: string | null;
  severity: string;
  summary: string;
  principal: string | null;
  method: string | null;
  resource: string | null;
  request_id: string | null;
  caller_ip: string | null;
  payload: Record<string, unknown>;
}

export interface LogsResponse {
  run_id: string;
  category: LogCategory;
  entries: LogEntry[];
  logging_enabled: boolean | null;
  data_access_notice?: boolean;
  setup_notice?: string | null;
}

const LOGGING = "https://logging.googleapis.com/v2";
const COMPUTE = "https://compute.googleapis.com/compute/v1";
const BEYONDCORP = "https://beyondcorp.googleapis.com/v1";
// Request only the fields needed for the operator-facing health/audit view.
// In particular, never receive textPayload, requestMetadata/callerIp, an
// authenticated principal, or Nginx URI/query fields. Chrome Web Store policy
// classifies URL paths obtained from a cloud service as web-browsing activity;
// this feature does not need that data.
const LOG_ENTRY_FIELDS = [
  "insertId",
  "timestamp",
  "severity",
  "protoPayload.methodName",
  "protoPayload.serviceName",
  "protoPayload.resourceName",
  "protoPayload.status",
  "jsonPayload.status",
  "jsonPayload.upstream_status",
  "jsonPayload.request_id",
  "jsonPayload.requestId",
].join(",");
const LOG_FIELDS = encodeURIComponent(`entries(${LOG_ENTRY_FIELDS}),nextPageToken`);
const MAX_LOG_PAGES = 100;

export class GatewayObservability {
  private readonly transport: Transport;

  constructor(transport: Transport) {
    this.transport = transport;
  }

  async listLogs(
    spec: DeploymentSpec,
    options: { runId: string; category: LogCategory; hours?: number; limit?: number },
  ): Promise<LogsResponse> {
    // Clamped rather than validated: an out-of-range value is a UI slip, not an
    // attack, and refusing the whole request would be less useful than serving
    // a sane window.
    const hours = Math.max(1, Math.min(options.hours ?? 24, 168));
    const limit = Math.max(1, Math.min(options.limit ?? 100, 200));
    const since = new Date(Date.now() - hours * 3_600_000).toISOString();
    const filters = [`timestamp>="${since}"`];
    let setupNotice: string | null = null;
    let dataAccessNotice = false;
    let loggingEnabled: boolean | null = null;

    if (options.category === "access") {
      filters.push(
        'resource.type="audited_resource"',
        'resource.labels.method="AuthorizeUser"',
        'resource.labels.service="beyondcorp.googleapis.com"',
      );
      // These are Data Access audit logs; they are off by default and the UI
      // says so rather than showing an empty list as if nothing happened.
      dataAccessNotice = true;
    } else if (options.category === "connection") {
      loggingEnabled = await this.gatewayLoggingEnabled(spec);
      filters.push('resource.type="beyondcorp.googleapis.com/SecurityGateway"');
    } else if (options.category === "admin") {
      filters.push(
        'resource.type="audited_resource"',
        'resource.labels.service="beyondcorp.googleapis.com"',
      );
    } else {
      const instanceId = await this.offloadInstanceId(spec);
      if (instanceId === null) {
        return {
          run_id: options.runId,
          category: options.category,
          entries: [],
          logging_enabled: loggingEnabled,
          setup_notice:
            spec.backend_kind === "direct_https"
              ? "Nginx logs require an HTTP-offload VM and the Google Cloud Ops Agent. " +
                "Direct HTTPS has no Nginx tier."
              : "The offload VM was not found or its instance ID is unavailable.",
        };
      }
      filters.push(
        'resource.type="gce_instance"',
        `resource.labels.instance_id="${instanceId}"`,
        '(log_id("nginx_access") OR log_id("nginx_access.log"))',
      );
      setupNotice =
        "Nginx entries appear after the Google Cloud Ops Agent is configured " +
        "to collect /var/log/nginx/sgstudio-access.log.";
    }

    const raw: Record<string, unknown>[] = [];
    const seenPageTokens = new Set<string>();
    let pageToken: string | undefined;
    for (let page = 0; page < MAX_LOG_PAGES; page += 1) {
      const { payload } = await this.transport.requestJson(
        "POST",
        `${LOGGING}/entries:list?fields=${LOG_FIELDS}`,
        {
          jsonBody: {
            resourceNames: [`projects/${spec.project_id}`],
            filter: filters.join(" AND "),
            orderBy: "timestamp desc",
            pageSize: Math.min(limit - raw.length, 200),
            ...(pageToken === undefined ? {} : { pageToken }),
          },
        },
      );
      const pageEntries = payload.entries === undefined ? [] : payload.entries;
      if (
        !Array.isArray(pageEntries) ||
        pageEntries.some((item) => item === null || typeof item !== "object" || Array.isArray(item))
      ) {
        throw new Error("cloud-logging-entries-invalid");
      }
      raw.push(
        ...(pageEntries as Record<string, unknown>[]).slice(0, limit - raw.length),
      );

      const next = payload.nextPageToken;
      if (next === undefined || next === "") break;
      if (typeof next !== "string" || seenPageTokens.has(next)) {
        throw new Error("cloud-logging-page-token-invalid");
      }
      seenPageTokens.add(next);
      if (raw.length >= limit) break;
      if (page + 1 >= MAX_LOG_PAGES) {
        throw new Error("cloud-logging-pagination-incomplete");
      }
      pageToken = next;
    }

    const entries = raw.map((item, index) => this.entry(item, options.category, index));

    return {
      run_id: options.runId,
      category: options.category,
      entries,
      logging_enabled: loggingEnabled,
      data_access_notice: dataAccessNotice,
      setup_notice: setupNotice,
    };
  }

  private async gatewayLoggingEnabled(spec: DeploymentSpec): Promise<boolean> {
    // `logging: {}` is the current Security Gateway connection-logging enable
    // marker. Request only that field so this state check cannot receive
    // gateway addresses or any unrelated provider output.
    const { payload } = await this.transport.requestJson(
      "GET",
      `${BEYONDCORP}/projects/${spec.project_id}/locations/global/` +
        `securityGateways/${spec.gateway_id}?fields=logging`,
    );
    if (!Object.prototype.hasOwnProperty.call(payload, "logging")) return false;
    const logging = payload.logging;
    if (
      logging === null ||
      typeof logging !== "object" ||
      Array.isArray(logging) ||
      Object.keys(logging as Record<string, unknown>).length !== 0
    ) {
      throw new Error("security-gateway-logging-state-invalid");
    }
    return true;
  }

  private async offloadInstanceId(spec: DeploymentSpec): Promise<string | null> {
    if (spec.backend_kind === "direct_https") return null;
    try {
      const { payload } = await this.transport.requestJson(
        "GET",
        `${COMPUTE}/projects/${spec.project_id}/zones/${spec.zone}/instances/${spec.name}-offload`,
      );
      return typeof payload.id === "string" ? payload.id : null;
    } catch {
      return null;
    }
  }

  /**
   * Reduce an entry to what the UI shows.
   *
   * Only explicitly allowlisted operational metadata leaves this function.
   * Payloads can carry request paths, query data, IP addresses, free text, and
   * principal identifiers; the views exist to answer "did an operation happen"
   * rather than to surface or persist those contents.
   */
  private entry(item: Record<string, unknown>, category: LogCategory, index: number): LogEntry {
    const proto = item.protoPayload as Record<string, unknown> | undefined;
    const json = item.jsonPayload as Record<string, unknown> | undefined;
    const status = json?.status ?? json?.upstream_status ??
      (proto?.status as Record<string, unknown> | undefined)?.code;
    const summary = String(proto?.methodName ?? proto?.serviceName ?? status ?? "audited event");
    return {
      insert_id: String(item.insertId ?? item.insert_id ?? `log-${index}`),
      category,
      timestamp: typeof item.timestamp === "string" ? item.timestamp : null,
      severity: String(item.severity ?? "DEFAULT"),
      summary: summary.slice(0, 500) || "(no summary)",
      principal: null,
      method: typeof proto?.methodName === "string" ? proto.methodName : null,
      resource: typeof proto?.resourceName === "string" ? proto.resourceName : null,
      request_id:
        typeof json?.request_id === "string"
          ? json.request_id
          : typeof json?.requestId === "string"
            ? json.requestId
            : null,
      caller_ip: null,
      payload: {
        method_name: proto?.methodName ?? null,
        service_name: proto?.serviceName ?? null,
        status: (json?.status ?? json?.upstream_status) ?? null,
      },
    };
  }
}
